### PR TITLE
PERF: pickling speed of rv_*

### PR DESCRIPTION
**Pull request by andyfaff**
This PR makes some changes to speed up the pickling of `rv_continuous`, `rv_generic`, `rv_discrete`, `rv_sampled` by:

- not calling `__init__` whenever a distribution is unpickled (in `rv_generic.__setstate__`). This saves calling `rv_generic.__init__`, `rv_continuous.__init__`. Those `__init__`'s do various checks (examining function signatures), create docstrings, etc.

- this means that the monkey patched methods created in the `rv_continuous.__init__` have to be done elsewhere. This is done by adding a `_construct_methods` to each of the subclasses of `rv_generic`.

- further optimizations would be possible with the `_random_state` attribute. Pickling/unpickling `np.random.RandomState` is expensive, pickling/unpickling `np.random.Generator` is a lot faster. It's not clear to me if the `_random_state` attribute is [deprecated](https://github.com/scipy/scipy/blob/master/scipy/stats/_distn_infrastructure.py#L986) (and following lines).

### CODE REVIEW

**Comment by andyfaff** [comment]
Before:
```
>>> import scipy.stats as stats
>>> import pickle
>>> %timeit stats.norm(scale=1, loc=1)
694 µs ± 123 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
>>> rv = stats.norm(scale=1, loc=1)
>>> %timeit s = pickle.dumps(rv); pickle.loads(s)
1.02 ms ± 24 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

After:
```
>>> import scipy.stats as stats
>>> import pickle
>>> %timeit stats.norm(scale=1, loc=1)
533 µs ± 12.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
>>> rv = stats.norm(scale=1, loc=1)
>>> %timeit s = pickle.dumps(rv); pickle.loads(s)
552 µs ± 12.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

What's more interesting to me is the reduced pickling time.

After these changes and using a `Generator`:
```
>>> rv = norm(scale=1, loc=1)
>>> rv.dist._random_state = np.random.default_rng()
>>> %timeit pkl = pickle.dumps(rv); pickle.loads(pkl)
332 µs ± 10.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Not having a `_random_state` attribute would save another ~100us.

**Comment by andyfaff** [comment]
There are a lot of test changes here, sorry. When the distribution instances are set up, `norm = norm_gen("norm")` they are automatically given a `_random_state` attribute that is set to the RandomState singleton in numpy, `np.random.mtrand._rand`. Thus, the following is `True` they are first initialised:

```
norm._random_state is np.random.mtrand._rand
```

However, in testing I don't necessarily think that behaviour is kept, and `_random_state` is not necessarily the same object as the `RandomState` singleton, but could vary depending on how/when the singleton or dist is setup by pytest. Therefore, relying on setting the state of the global RandomState instance is a bad idea because it won't give a reproducible test. This could be the source of a lot of flaky tests.
Therefore, if `dist.rvs` relies on reproducible random number generation the `dist.rvs` method should be supplied an explicit `random_state` argument.

**Comment by andyfaff** [comment]
I have an idea for further unpickling speedup. Whenever the objects are unpickled the `_parse_args`, `_parse_args_rvs`, `_parse_args_stats` functions have to be dynamically recreated. This takes quite a bit of time. It should be possible to create a class that accepts the `shape_list` and `locscale_in`, `locscale_out`, and has methods to provide those parsing functions.

e.g.
```
class _ArgParser(object):
    def __init__(self, shapes, locscale_in, locscale_out):
        pass

    def _parse_args(self, *args, **kwds):
        # called by rv_generic subclasses
```
Objects of this class would be easily picklable (and speedy).

**Comment by andyfaff** [comment]
I should emphasise that these changes would have the effect of invalidating **all** `rv_continuous`/`rv_discrete`/`rv_sample` pickles made using an older version of scipy. It would be possible to retain the back compatibility at the expense of cruft.

I'm not sure what our policy is on the back compatibility of pickles.

**Comment by andyfaff** [comment]
With the last commit:
```
>>> rv = norm(scale=1, loc=1)
>>> rv.dist._random_state = np.random.default_rng()
>>> %timeit pkl = pickle.dumps(rv); pickle.loads(pkl)
240 µs ± 12.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
This drops to 130us if `_random_state` is cleared before pickling.

**Comment by pvanmulbregt** [comment]
The ` _random_state` member itself isn't deprecated, just accessing it directly from inside any subclass's `_rvs()` method. That was needed for thread safety for callers pass in n a random state.
`_random_state` does store the state which gets used in `rvs()` if the caller doesn't pass one in. That maybeven be the most common used case.

File: scipy/stats/tests/test_continuous_basic.py [hunk 1]
```diff
@@ -106,13 +107,13 @@ def test_cont_basic(distname, arg):
...
         distname = 'rv_histogram_instance'
-    np.random.seed(765456)
+    rng = np.random.RandomState(765456)
     sn = 500
     with npt.suppress_warnings() as sup:
         # frechet_l and frechet_r are deprecated, so all their
         # methods generate DeprecationWarnings.
         sup.filter(category=DeprecationWarning, message=".*frechet_")
-        rvs = distfn.rvs(size=sn, *arg)
+        rvs = distfn.rvs(size=sn, *arg, random_state=rng)
```
Comment (ev-br) [inline scipy/stats/tests/test_continuous_basic.py:116]: Are these changes (`random_state = ...` instead of `np.random.seed`) necessary for the tests to pass or is it a cleanup?

If it's a cleanup, the cleanup's good but maybe better separate them to their own PR.  If they are needed, then why --- np.random.seed was OK before and is no longer OK with this PR?

File: scipy/stats/tests/test_continuous_basic.py [hunk 2]
```diff
@@ -662,3 +664,40 @@ def test_methods_with_lists(method, distname, args):
...
+
+@pytest.mark.parametrize('distname, args', distcont)
+def test_pickle_frozen(distname, args):
+    # check that you can pickle frozen distributions
+    with npt.suppress_warnings() as sup:
+        sup.filter(category=DeprecationWarning, message=".*frechet_")
+        dist = getattr(stats, distname)
+
+        x = np.array([0.1, 0.2])
+        shape2 = [[a]*2 for a in args]
```
Comment (ev-br) [inline scipy/stats/tests/test_continuous_basic.py:676]: shape2 not used?

File: scipy/stats/tests/test_continuous_basic.py [hunk 3]
```diff
@@ -662,3 +664,40 @@ def test_methods_with_lists(method, distname, args):
...
+    # first unfrozen
+    norm_gen = stats.norm
+    norm_gen.random_state = 4323
+    pkl = pickle.dumps(norm_gen)
+    norm_gen2 = pickle.loads(pkl)
+
+    samples = norm_gen.rvs(size=5, scale=2, loc=-1)
+    npt.assert_allclose(norm_gen2.rvs(size=5, scale=2, loc=-1), samples)
+
+    # then unfrozen distribution
```
Comment (ev-br) [inline scipy/stats/tests/test_continuous_basic.py:698]: typo? first unfrozen ... then frozen?

File: scipy/stats/_distn_infrastructure.py [hunk 4]
```diff
@@ -618,21 +618,24 @@ def random_state(self):
...
+        Generates the argument-parsing functions dynamically and attaches
+        them to the instance.
 
-    def __setstate__(self, state):
-        ctor_param, r = state
-        self.__init__(**ctor_param)
-        self._random_state = r
-        return self
+        Should be called after _construct_argparser, and during unpickling
+        (__setstate__)
```
Comment (ev-br) [inline scipy/stats/_distn_infrastructure.py:627]: Does introducing this method change recommendations for subclassers? ISTM it's all hidden as long as `rv_generic.__init__` is called, but I'd rather ask to re-confirm.

File: scipy/stats/_distn_infrastructure.py [hunk 5]
```diff
@@ -1660,6 +1648,42 @@ def __init__(self, momtype=1, a=None, b=None, xtol=1e-14,
...
+        # these methods will be remade in __setstate__
+        # _random_state attribute is taken care of by rv_generic
+        attrs = ["_parse_args", "_parse_args_stats", "_parse_args_rvs",
+                 "_cdfvec", "_ppfvec", "vecentropy", "generic_moment"]
+        [dct.pop(attr, None) for attr in attrs]
+        return dct
+
+    def __setstate__(self, state):
+        dct = state
+        self.__dict__.update(dct)
```
Comment (ev-br) [inline scipy/stats/_distn_infrastructure.py:1663]: is `dct` a leftover? cf similar stanzas in `rv_discete`

File: scipy/stats/_distn_infrastructure.py [hunk 6]
```diff
@@ -2879,6 +2901,29 @@ def __init__(self, a=0, b=inf, name=None, badvalue=None,
...
+        dct = self.__dict__.copy()
+        # these methods will be remade in __setstate__
+        attrs = ["_parse_args", "_parse_args_stats", "_parse_args_rvs",
+                 "_cdfvec", "_ppfvec", "generic_moment"]
+        [dct.pop(attr, None) for attr in attrs]
+        return dct
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        self._attach_methods()
```
Comment (ev-br) [inline scipy/stats/_distn_infrastructure.py:2917]: Actually, can `__get/setstate__` live in `rv_generic` instead of being duplicated in `rv_continuous` and `rv_discrete`?

EDIT. I see, one has `vecentropy` and the other one does not. Why? Anyway, can be left for a follow-up PR.

File: scipy/stats/_distn_infrastructure.py [hunk 7]
```diff
@@ -2894,8 +2939,6 @@ def __init__(self, a=0, b=inf, name=None, badvalue=None,
         # now that self.numargs is defined, we can adjust nin
         self._cdfvec.nin = self.numargs + 1
 
-        self._construct_docstrings(name, longname, extradoc)
```
Comment (ev-br) [inline scipy/stats/_distn_infrastructure.py:2897]: why removing it?

File: scipy/stats/_distn_infrastructure.py [hunk 8]
```diff
@@ -3538,13 +3581,30 @@ def __init__(self, a=0, b=inf, name=None, badvalue=None,
...
+        # these methods will be remade in __setstate__
+        # __setstate__ is supplied by rv_discrete
+        attrs = ["_parse_args", "_parse_args_stats", "_parse_args_rvs"]
+        [dct.pop(attr, None) for attr in attrs]
+
+        return dct
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        self._attach_argparser_methods()
```
Comment (ev-br) [inline scipy/stats/_distn_infrastructure.py:3606]: The comment above says it's supplied by rv_discrete, so no needed here (or the comment should be fixed)

File: scipy/stats/tests/test_continuous_basic.py [hunk 9]
```diff
@@ -662,3 +664,40 @@ def test_methods_with_lists(method, distname, args):
                             [f(*v) for v in zip(x, *shape2, loc, scale)],
                             rtol=1e-15, atol=1e-15)
 
+
+@pytest.mark.parametrize('distname, args', distcont)
+def test_pickle_frozen(distname, args):
```
Comment (ev-br) [inline scipy/stats/tests/test_continuous_basic.py:669]: These tests are good and seem to be more extensive then  `test_pickling` at https://github.com/scipy/scipy/blob/master/scipy/stats/tests/test_distributions.py#L2752
It's be also good to consolidate and remove duplication.

File: scipy/stats/tests/test_continuous_basic.py [hunk 10]
```diff
@@ -662,3 +664,40 @@ def test_methods_with_lists(method, distname, args):
                             [f(*v) for v in zip(x, *shape2, loc, scale)],
                             rtol=1e-15, atol=1e-15)
 
+
+@pytest.mark.parametrize('distname, args', distcont)
```
Comment (ev-br) [inline scipy/stats/tests/test_continuous_basic.py:668]: So only continuous distributions are tested? Since there are similar code paths in `rv_discrete`, these should be tested, too.

**Review by ev-br — COMMENTED** [review]
Looks reasonable to me. Left a few comments inline, mostly minor stuff.

Bigger questions:
1. Invalidation of old pickles. This should certainly be mentioned in the release notes. It is probably OK, but I actually don't know if it's a big deal for somebody or not. Mention it in the scipy-dev/scipy-user mailing lists?

2.  Impact on downstream. Ideally, this all should be internal to scipy.stats, but many details leaked. statsmodels is a big user, are you in position to check if statsmodels is affected by this PR? (I am not ATM, sadly). 

And it'd be good to merge this early in the release cycle, for the same reasons: hard to predict who's affected.

**Comment by ev-br** [comment]
W.r.t. backcompat break with older pickles: maybe it's easier to just do https://github.com/scipy/scipy/pull/12550#issuecomment-659332983 and forget :-)

**Comment by andyfaff** [comment]
I can check with statsmodels to see if the changes are an issue. I guess the biggest change is that `__init__` is not called after every unpickle (the stuff in __init__ is expensive). If there are downstream subclasses of `rv_generic`, `rv_continuous`, `rv_generic`, that rely on the superclass behaviour, then they might need a little adjustment. I think maintaining pickle compatibility should be possible with my previous suggestion.

File: scipy/stats/tests/test_continuous_basic.py (hunk 1 above)
Comment (andyfaff) [reply scipy/stats/tests/test_continuous_basic.py:116]: When the distributions are first created on import the following is True:
```
from scipy.stats import norm
assert norm._random_state is np.random.mtrand._rand
```
i.e. the `_random_state` attribute is the numpy global RandomState singleton object. Thus setting the random seed via `np.random.seed` affects the random number stream from `norm.rvs`. However, as I was making these changes I found that various tests were failing, those that relied on a specific output from random number generation. I then put the assert given above into the failing tests, and the assert failed. I therefore suspect that in some cases (possibly a pytest side effect), the numpy RandomState singleton is not the same object as that contained in all the distributions. I don't know why that is. That being the case, using `np.random.seed` would have no effect on the stream of `distfn.rvs`.
I made changes like the line this comment is attached to on the failing tests, and I also decided to make similar changes to the test files they contained. I have seen some of the same stats tests being flaky from time to time, this may be the cause.

File: scipy/stats/tests/test_continuous_basic.py (hunk 2 above)
Comment (andyfaff) [reply scipy/stats/tests/test_continuous_basic.py:676]: Not used, I'll remove

File: scipy/stats/tests/test_continuous_basic.py (hunk 3 above)
Comment (andyfaff) [reply scipy/stats/tests/test_continuous_basic.py:698]: typo, fixed.

File: scipy/stats/_distn_infrastructure.py (hunk 5 above)
Comment (andyfaff) [reply scipy/stats/_distn_infrastructure.py:1663]: changed

File: scipy/stats/_distn_infrastructure.py (hunk 7 above)
Comment (andyfaff) [reply scipy/stats/_distn_infrastructure.py:2897]: It's not actually removed, it's the diff that looks funny. If you look at the entire file `_construct_docstrings` is still there.

File: scipy/stats/tests/test_continuous_basic.py (hunk 9 above)
Comment (andyfaff) [reply scipy/stats/tests/test_continuous_basic.py:669]: My function is probably duplicating [check_pickling in common_tests](https://github.com/scipy/scipy/blob/master/scipy/stats/tests/common_tests.py#L273). The stats test suite is so large that sometimes it's difficult to know what's out there, it's so easy to overlook something. I might remove the test I added, I'm not sure it is better than that in common_tests, and the original test also tests discrete.

File: scipy/stats/tests/test_continuous_basic.py (hunk 10 above)
Comment (andyfaff) [reply scipy/stats/tests/test_continuous_basic.py:668]: it seems that `check_pickling` [is also used for testing rv_discrete](https://github.com/scipy/scipy/blob/master/scipy/stats/tests/test_discrete_basic.py#L57).

File: scipy/stats/tests/test_continuous_basic.py (hunk 9 above)
Comment (andyfaff) [reply scipy/stats/tests/test_continuous_basic.py:669]: What check_pickling doesn't do is test if the frozen distribution pickle/unpickles. The link you provided in [test_distributions](https://github.com/scipy/scipy/blob/master/scipy/stats/tests/test_distributions.py#L2752) does do that, but only for 3 distributions. I might amend `check_pickling` to test them all.

File: scipy/stats/_distn_infrastructure.py (hunk 4 above)
Comment (andyfaff) [reply scipy/stats/_distn_infrastructure.py:627]: If subclassers call `rv_generic.__init__` (or if they subclass `rv_continuous`, then `rv_continuous.__init__`) everything should be ok. However, if they have special unpickling behaviour, overriding `__setstate__`, or expecting `__init__` to be called from `__setstate__`, then they may experience problems.

File: scipy/stats/_distn_infrastructure.py (hunk 6 above)
Comment (andyfaff) [reply scipy/stats/_distn_infrastructure.py]: The dynamically created methods `_parse_args`, `_parse_args_stats`, `_parse_args_rvs` need to be removed from all the state dictionaries otherwise there is an exception raised in `__setstate__` when `self.__dict__.update(state)` is called. In addition, rv_continuous has methods like `_cdfvec`  that also need to be removed for the same reason. Thus, each class has it's own `__getstate__`. However, in commit a3ea650cd6240aad3ed321766b7253ac17ecc6a8 I've made `rv_generic.__setstate__` common to all subclasses. This means each class needs to implement an `_attach_methods` method.

File: scipy/stats/_distn_infrastructure.py (hunk 8 above)
Comment (andyfaff) [reply scipy/stats/_distn_infrastructure.py:3629]: Changed in a3ea650cd6240aad3ed321766b7253ac17ecc6a8

File: scipy/stats/tests/test_continuous_basic.py (hunk 9 above)
Comment (andyfaff) [reply scipy/stats/tests/test_continuous_basic.py]: By a3ea650cd6240aad3ed321766b7253ac17ecc6a8 check_pickling also tests pickling of all frozen distributions.

**Comment by andyfaff** [comment]
@ev-br, thanks for the quick review. I've made the changes you suggested.

**Comment by andyfaff** [comment]
ping @josef-pkt, @bashtage this PR attempts to speedup pickling of rv_generic and rv_frozen instances. This can be a hurdle when using various distributions in a multiprocessing context.

It mainly achieves this speedup by avoiding having to call `self.__init__` in `__setstate__` as the unpickling process unfurls. The various `__init__`s have lots of time intensive things, like looking at function signatures, constructing docstrings, etc

As such `__setstate__` and `__getstate__` are modified by a fair amount, and each class gets it's own `_attach_methods` method that is responsible for attaching dynamic methods.

Since statsmodels is probably the biggest consumer of `rv_*`, I was wondering if you could give this PR a look over and let me know if statsmodels would be adversely affected by the changes made here.

**Comment by ev-br** [comment]
Needs a rebase now.

**Comment by andyfaff** [comment]
@ev-br, that was not the easiest rebase.

**Comment by ev-br** [comment]
All seems to be backwards compatible, no further objections, so merging. Thanks @andyfaff , nice improvement!
